lib/shelf_io.dart

```diff
@@ -74,8 +74,14 @@ Future handleRequest(HttpRequest request, Handler handler) async {
   try {
     shelfRequest = _fromHttpRequest(request);
   } catch (error, stackTrace) {
-    var response =
-        _logTopLevelError('Error parsing request.\n$error', stackTrace);
+    final isClientError = error is ArgumentError &&
+        (error.name == 'method' || error.name == 'requestedUri');
+    _logTopLevelError('Error parsing request.\n$error', stackTrace);
+    final response = isClientError
+        ? Response(400,
+            body: 'Bad Request',
+            headers: {HttpHeaders.contentTypeHeader: 'text/plain'})
+        : Response.internalServerError();
```

Comment:
Maybe we should unwrap this a bit...
```dart
try {
} on ArgumentError catch (error) {
  if (error.name == 'method' || error.name == 'requestedUri') {
    .... 400
   return;
  }
  .... 500;
  return;
} catch (error, stackTrace) {
  var response =
        _logTopLevelError('Error parsing request.\n$error', stackTrace);
  await _writeResponse(response, request.response);
} 
```
But catching an `Error` subclass is almost always wrong. The correct thing to do would be to duplicate the checks... As we can't throw a different `Exception`, since `ArgumentError` is the correct thing to throw if invariants are violated in a public constructor. This is nits... 

But doing `} on ArgumentError catch (error) {` would clean up this a lot...

Reply:
Updated the catch block. I'm not convinced that it adds enough value to warrant it, but I agree, it is a more correct version of the same code.